src/pybind/mgr/dashboard/controllers/docs.py

```diff
@@ -0,0 +1,298 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import cherrypy
+
+from . import Controller, BaseController, AuthRequired, Endpoint, ENDPOINT_MAP
+from .. import logger
+
+
+@Controller('docs')
+@AuthRequired()
+class Docs(BaseController):
+
+    @classmethod
+    def _gen_tags(cls, all_endpoints):
+        ctrl_names = set()
+        for endpoints in ENDPOINT_MAP.values():
+            for endpoint in endpoints:
+                if endpoint.is_api or all_endpoints:
+                    ctrl_names.add(endpoint.group)
+
+        tags = ""
+        for name in sorted(ctrl_names):
+            tags += """
+            {{ "name": "{}", "description": ""}},
+            """.format(name)
+
+        return tags[:-1]
+
+    @classmethod
+    def _gen_type(cls, param):
+        # pylint: disable=too-many-return-statements
+        """
+        Generates the type of parameter based on its name and default value,
+        using very simple heuristics.
+        """
+        param_name = param['name']
+        def_value = param['default'] if 'default' in param else None
+        if param_name.startswith("is_"):
+            return "boolean"
+        elif "size" in param_name:
+            return "integer"
+        elif "count" in param_name:
+            return "integer"
+        elif "num" in param_name:
+            return "integer"
+        elif isinstance(def_value, bool):
+            return "boolean"
+        elif isinstance(def_value, int):
+            return "integer"
+        return "string"
+
+    @classmethod
+    def _format_val(cls, val):
+        if isinstance(val, bool):
+            return "true" if val else "false"
+        elif isinstance(val, int):
+            return val
+        elif val is None:
+            return "null"
+        return '"{}"'.format(val)
+
+    @classmethod
+    def _gen_body_param(cls, body_params):
+        required = [p['name'] for p in body_params if p['required']]
+        required = ", ".join(['"{}"'.format(p) for p in required])
+
+        props = []
+        for p in body_params:
+            param_str = '"{}": {{"type": "{}"'.format(p['name'],
+                                                      cls._gen_type(p))
+            if 'default' in p:
+                param_str += ', "default": {}'.format(
+                    cls._format_val(p['default']))
+            param_str += "}"
+            props.append(param_str)
+        props = ", ".join(props)
+        if not props:
+            return None
+
+        return """{{
+            "in": "body",
+            "name": "body",
+            description: "",
+            "required": true,
+            "schema": {{
+                "type": "object",
+                "required": [{}],
+                "properties": {{
+                    {}
+                }}
+            }}
+        }}""".format(required, props)
+
+    @classmethod
+    def _gen_responses_descriptions(cls, method):
+        resps = []
```

Comment:
I suggest to use a dict which is finally converted to JSON. Thus you do not have to take care that you create valid JSON strings.